services/worker-manager/src/providers/aws.js

```diff
@@ -82,83 +90,90 @@ class AwsProvider extends Provider {
       });
     }
 
-    // Make sure we don't get "The same resource type may not be specified more than once in tag specifications" errors
-    const TagSpecifications = config.launchConfig.TagSpecifications ? config.launchConfig.TagSpecifications : [];
-    const instanceTags = [];
-    const otherTagSpecs = [];
-    TagSpecifications.forEach(ts =>
-      ts.ResourceType === 'instance' ? instanceTags.concat(ts.Tags) : otherTagSpecs.push(ts)
-    );
+    const shuffledConfigs = _.shuffle(workerPool.config.launchConfigs);
 
     let spawned;
-    try {
-      spawned = await ec2.runInstances({
-        ...config.launchConfig,
-
-        UserData: userData.toString('base64'), // The string needs to be base64-encoded. See the docs above
-
-        MaxCount: toSpawn,
-        MinCount: toSpawn,
-        TagSpecifications: [
-          ...otherTagSpecs,
-          {
-            ResourceType: 'instance',
-            Tags: [
-              ...instanceTags,
-              {
-                Key: 'CreatedBy',
-                Value: `taskcluster-wm-${this.providerId}`,
-              }, {
-                Key: 'Owner',
-                Value: workerPool.owner,
-              },
-              {
-                Key: 'ManagedBy',
-                Value: 'taskcluster',
-              },
-              {
-                Key: 'Name',
-                Value: `${workerPoolId}`,
-              }],
-          },
-        ],
-      }).promise();
-    } catch (e) {
-      this.monitor.err(`Error calling AWS API: ${e}`);
+    for await (let config of shuffledConfigs) {
+      // Make sure we don't get "The same resource type may not be specified
+      // more than once in tag specifications" errors
+      const TagSpecifications = config.TagSpecifications ? config.TagSpecifications : [];
+      const instanceTags = [];
+      const otherTagSpecs = [];
+      TagSpecifications.forEach(ts =>
+        ts.ResourceType === 'instance' ? instanceTags.concat(ts.Tags) : otherTagSpecs.push(ts)
+      );
+
+      try {
+        spawned = await this.ec2s[config.region].runInstances({
+          ...config.launchConfig,
+
+          UserData: userData.toString('base64'), // The string needs to be base64-encoded. See the docs above
+
+          MaxCount: toSpawn,
+          MinCount: toSpawn,
```

Comment:
I think the only issue here is that if estimator asks for 4 instances and the user has 3 `launchConfigs` specified, this will spawn 12 instances, right? In my head the way this would work would be looping over `shuffledConfigs` spawning at most N instances at a time until we've spawned `toSpawn` instances. Does that make sense or am I missing something.

Reply:
ah yes yes